Compiling NIO in Release Mode throws warnings

### Expected behavior
No warnings, clean compile.

### Actual behavior
A set of warnings, no clean compile.

```
Compile Swift Module 'NIO' (47 sources)
/Users/helge/dev/Swift/SwiftXcode/Images/SwiftNIO_XcodeImage/.build/checkouts/swift-nio.git-3108475404973543938/Sources/NIO/EventLoopFuture.swift:316:25: warning: will never be executed
            if let me = eventLoop as? SelectableEventLoop {
                        ^
/Users/helge/dev/Swift/SwiftXcode/Images/SwiftNIO_XcodeImage/.build/checkouts/swift-nio.git-3108475404973543938/Sources/NIO/EventLoopFuture.swift:315:12: note: condition always evaluates to false
        if _isDebugAssertConfiguration() {
           ^
/Users/helge/dev/Swift/SwiftXcode/Images/SwiftNIO_XcodeImage/.build/checkouts/swift-nio.git-3108475404973543938/Sources/NIO/EventLoopFuture.swift:338:64: warning: will never be executed
        if _isDebugAssertConfiguration(), let eventLoop = self.eventLoop as? SelectableEventLoop {
                                                               ^
/Users/helge/dev/Swift/SwiftXcode/Images/SwiftNIO_XcodeImage/.build/checkouts/swift-nio.git-3108475404973543938/Sources/NIO/EventLoopFuture.swift:338:12: note: condition always evaluates to false
        if _isDebugAssertConfiguration(), let eventLoop = self.eventLoop as? SelectableEventLoop {
           ^
Compile Swift Module 'NIOTLS' (3 sources)
Compile Swift Module 'NIOHTTP1' (7 sources)
Compile Swift Module 'NIOOpenSSL' (11 sources)
Compile Swift Module 'MicroExpress' (7 sources)
/Users/helge/dev/Swift/SwiftXcode/Images/SwiftNIO_XcodeImage/.build/checkouts/MicroExpress.git-6340581023861930988/Sources/MicroExpress/Express.swift:23:26: warning: 'addHTTPServerHandlers(first:)' is deprecated: Please use configureHTTPServerPipeline
        channel.pipeline.addHTTPServerHandlers().then {
                         ^
```

### Steps to reproduce

`swift build -c release`

### SwiftNIO version/commit hash

1.2.1

### Swift & OS version (output of `swift --version && uname -a`)

Apple Swift version 4.0.3 (swiftlang-900.0.74.1 clang-900.0.39.2)
Target: x86_64-apple-macosx10.9
Darwin ZeaPro.local 16.7.0 Darwin Kernel Version 16.7.0: Thu Jan 11 22:59:40 PST 2018; root:xnu-3789.73.8~1/RELEASE_X86_64 x86_64

--- a/Sources/NIO/EventLoopFuture.swift
+++ b/Sources/NIO/EventLoopFuture.swift
@@ -312,7 +312,7 @@ public final class EventLoopFuture<T> {
         self.value = value
         self._fulfilled = Atomic(value: value != nil)
 
-        if _isDebugAssertConfiguration() {
+        debugOnly {
             if let me = eventLoop as? SelectableEventLoop {
                 me.promiseCreationStoreAdd(future: self, file: file, line: line)
             }
@@ -335,13 +335,15 @@ public final class EventLoopFuture<T> {
     }
 
     deinit {
-        if _isDebugAssertConfiguration(), let eventLoop = self.eventLoop as? SelectableEventLoop {
-            let creation = eventLoop.promiseCreationStoreRemove(future: self)
-            if !fulfilled {
-                fatalError("leaking promise created at \(creation)", file: creation.file, line: creation.line)
+        debugOnly {
+            if let eventLoop = self.eventLoop as? SelectableEventLoop {
+                let creation = eventLoop.promiseCreationStoreRemove(future: self)
+                if !fulfilled {
+                    fatalError("leaking promise created at \(creation)", file: creation.file, line: creation.line)
+                }
+            } else {
+                precondition(fulfilled, "leaking an unfulfilled Promise")
             }
-        } else {
-            precondition(fulfilled, "leaking an unfulfilled Promise")
         }
     }
 }

--- a/Sources/NIO/Utilities.swift
+++ b/Sources/NIO/Utilities.swift
@@ -28,6 +28,15 @@ private func darwinCoreCount() -> Int {
 }
 #endif
 
+/// A utility function that runs the body code only in debug builds, without
+/// emitting compiler warnings.
+///
+/// This is currently the only way to do this in Swift: see
+/// https://forums.swift.org/t/support-debug-only-code/11037 for a discussion.
+internal func debugOnly(_ body: () -> Void) {
+    assert({ body(); return true }())
+}
+
 /// Allows to "box" another value.
 final class Box<T> {
     let value: T

